src/jit/codegenxarch.cpp

```diff
@@ -4607,79 +4607,87 @@ void CodeGen::genCodeForIndexAddr(GenTreeIndexAddr* node)
+    const regNumber dstReg   = node->gtRegNum;
 
     // NOTE: `genConsumeReg` marks the consumed register as not a GC pointer, as it assumes that the input registers
     // die at the first instruction generated by the node. This is not the case for `INDEX_ADDR`, however, as the
     // base register is multiply-used. As such, we need to mark the base register as containing a GC pointer until
     // we are finished generating the code for this node.
 
-    gcInfo.gcMarkRegPtrVal(base->gtRegNum, base->TypeGet());
-    assert(!varTypeIsGC(index->TypeGet()));
+    gcInfo.gcMarkRegPtrVal(baseReg, base->TypeGet());
+    assert(varTypeIsIntegral(index->TypeGet()));
 
     regNumber tmpReg = REG_NA;
 
     // Generate the bounds check if necessary.
     if ((node->gtFlags & GTF_INX_RNGCHK) != 0)
     {
-        // Create a GT_IND(GT_LEA)) tree for the array length access.
-        GenTreeAddrMode arrLenAddr(base->TypeGet(), base, nullptr, 0, node->gtLenOffset);
-        arrLenAddr.gtRegNum = REG_NA;
-        arrLenAddr.SetContained();
-
-        GenTreeIndir arrLen = indirForm(TYP_INT, &arrLenAddr);
-
 #ifdef _TARGET_64BIT_
         // The CLI Spec allows an array to be indexed by either an int32 or a native int.  In the case that the index
-        // is a native int on a 64-bit platform, we will need to widen the array length and the compare.
+        // is a native int on a 64-bit platform, we will need to widen the array length and then compare.
         if (index->TypeGet() == TYP_I_IMPL)
         {
-            // Load the array length into a register.
-            tmpReg          = node->GetSingleTempReg();
-            arrLen.gtRegNum = tmpReg;
-            arrLen.ClearContained();
-            getEmitter()->emitInsLoadInd(ins_Load(TYP_INT), EA_4BYTE, arrLen.gtRegNum, &arrLen);
+            tmpReg = node->GetSingleTempReg();
+            getEmitter()->emitIns_R_AR(INS_mov, EA_4BYTE, tmpReg, baseReg, static_cast<int>(node->gtLenOffset));
+            getEmitter()->emitIns_R_R(INS_cmp, EA_8BYTE, indexReg, tmpReg);
         }
         else
-#endif
+#endif // _TARGET_64BIT_
         {
-            assert(varTypeIsIntegral(index->TypeGet()));
-
-            arrLen.gtRegNum = REG_NA;
-            arrLen.SetContained();
+            getEmitter()->emitIns_R_AR(INS_cmp, EA_4BYTE, indexReg, baseReg, static_cast<int>(node->gtLenOffset));
         }
 
-        // Generate the range check.
-        getEmitter()->emitInsBinary(INS_cmp, emitTypeSize(TYP_I_IMPL), index, &arrLen);
         genJumpToThrowHlpBlk(EJ_jae, SCK_RNGCHK_FAIL, node->gtIndRngFailBB);
     }
 
+#ifdef _TARGET_64BIT_
+    if (index->TypeGet() != TYP_I_IMPL)
+    {
+        // LEA needs 64-bit operands so we need to widen the index if it's TYP_INT.
+        // Since it's TYP_INT the upper 32 bits aren't used so we should be able
+        // to widen in place, without needing a temporary register.
+        getEmitter()->emitIns_R_R(INS_mov, EA_4BYTE, indexReg, indexReg);
```

Comment:
I'm a bit concerned about this. Sure, it's a `TYP_INT` value so the upper 32 bits should not be used.

But what if LSRA in the future does something like assign a 64 bit register variable to a 32 bit use - `a[(int)longVarReg]`  - and this code ends up zeroing out the upper 32 bits of `longVarReg`.

@CarolEidt  Opinions ?

Reply:
Ah, very good point. I'm believe that today that would be a cast and so we would only reuse the register if the source was a last use (and `index` would be a cast). To be conservative, one could assert that `index` is not a register candidate lclVar, or it is a `TYP_INT`, with a comment that, in future, if we make the cast "implicit" for this case, we need to use a temp if the lclVar is not a last use.

Reply:
Looks like this kind of register "reinterpretation" already occurs, but in a different manner:
```
N014 (  2,  3) [000007] ------------         t7 = *  CAST      long <- int REG rdx
                                                 /--*  t5     long   
                                                 +--*  t7     long   
N016 ( 10, 11) [000008] ---X--------         t8 = *  ADD       long   REG rax
                                                 /--*  t8     long   
N018 ( 14, 14) [000010] DA-X--------              *  STORE_LCL_VAR long   V02 tmp1          NA REG NA
N020 (  3,  4) [000018] -c--G--N----        t18 =    CLS_VAR_ADDR byref  Hnd=0xe8001558 REG NA
; V02 is long but here it is used as int and of course, the use gets the same register
N022 (  3,  2) [000013] C-----------        t13 =    LCL_VAR   int    V02 tmp1          rax REG rax
N024 (  1,  1) [000014] -c----------        t14 =    CNS_INT   int    3 REG NA
                                                 /--*  t13    int    
                                                 +--*  t14    int    
N026 (  8,  6) [000016] ------------        t16 = *  MUL       int    REG rax
```
This happens even in minopts so it looks like I'll need to actually allocate and the temp register when the array index needs to be widened, it cannot be done in place.

Reply:
Thanks for investigating this. This is rather troubling, but probably too fundamental to be easily changed.

Reply:
Yes, IMO it would better to not have LCLVAR nodes with a type other than the variable itself but I'm not sure if it's possible to avoid all the CQ fallout from adding the required cast nodes. I would guess it should be possible to make those cast nodes contained, though it's probably a bit complicated.

But what bugs me the most is that this kind of reinterpretation occurs even in minopts mode.